### PR TITLE
change the port configuration for fwd and some struct modify.

### PR DESCRIPTION
**Pull request by BillZhou734**
Signed-off-by: Dong Zhou <dongzhou@nvidia.com>

### CODE REVIEW

File: examples/vnet/doca_gw.c [hunk 1]
```diff
@@ -101,31 +84,29 @@ int doca_gw_rm_entry(uint16_t pipe_queue, struct doca_gw_pipelne_entry *entry)
  *
  * @return 
  */
-struct doca_gw_port * doca_gw_port_start(struct doca_gw_port_cfg *cfg, struct doca_gw_error *err)
+struct doca_gw_port * doca_gw_port_start(struct doca_gw_port *port, struct doca_gw_error *err)
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.c]: This is user API, if user need to provide the port struct then it means you expose implementation. it blocks from having any implementation in the future. please remain with CFG and PORT

File: examples/vnet/doca_gw.c [hunk 2]
```diff
@@ -101,31 +84,29 @@ int doca_gw_rm_entry(uint16_t pipe_queue, struct doca_gw_pipelne_entry *entry)
...
+	struct doca_gw_port_cfg *cfg = &port->cfg;
+
+	switch(cfg->type) {
+		case DOCA_GW_PORT_DPDK:
+			DOCA_LOG_INFO("port is dpdk, matching port id");
+			// init all required data sturcures for port.
+			break;
+		case DOCA_GW_PORT_DPDK_BY_ID:
+			DOCA_LOG_INFO("new doca port type:dpdk port id:%s", cfg->devargs);
+			ret = doca_gw_dpdk_start_port(port->port_id, cfg, (struct doca_dpdk_port*)port->user_data);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.c:122]: cfg->port

File: examples/vnet/doca_gw.c (hunk 1 above)
Comment (rbaryanai) [reply examples/vnet/doca_gw.c]: use this API:
struct doca_gw_port * doca_gw_port_start(struct doca_gw_port_cfg *cfg, struct doca_gw_error *err)

File: examples/vnet/doca_gw.h [hunk 3]
```diff
@@ -76,6 +76,46 @@ struct doca_gw_error {
...
+
+/**
+ * @brief - forwarding configuration
+ */
+struct doca_fwd_table_cfg {
+    enum doca_fwd_tbl_type type;
+    union {
+        struct fwd_rss {
+            uint32_t rss_flags;
+            uint16_t queues[MAX_RSS_QUEUE]; // reduce malloc ?
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.h]: why? this is a cfg object it should be done once in most cases.
putting a restriction here is not good. 16 might be too small

File: examples/vnet/doca_gw.h [hunk 4]
```diff
@@ -76,6 +76,46 @@ struct doca_gw_error {
...
+            int      num_queues;
+        } rss;
+        struct queue {
+            int queue_id;
+        } queue;
+        struct port {
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.h]: can you use line space here between different logical sturcutres, it is more readable

File: examples/vnet/gw.c [hunk 5]
```diff
@@ -467,14 +475,36 @@ static int gw_create(void)
...
+/*same as rte_eth_conf port_conf*/
+static struct gw_port_cfg cfg_port0 = {
+	.port_id = 0,
+	.def_fwd = {
+		.type = DOCA_FWD_QUEUE,
+		.queue = { .queue_id = 0},
+	},
+	.offload_fwd = {
+		.type = DOCA_FWD_RSS,
+		.rss = { .rss_flags = DOCA_RSS_IP, .queues = {0,3,4},.num_queues = 3,},
```
Comment (rbaryanai) [inline examples/vnet/gw.c]: i don't understand why RSS is a port parameter. I can have two different RSS actions for same port right? one using port 01 and second 2 3, or 0123 ...etc. so this should not be part of the port. RSS if FWD configuration.

File: examples/vnet/doca_gw_dpdk.h [hunk 6]
```diff
@@ -269,6 +279,15 @@ doca_gw_dpdk_pipe_create_flow(struct doca_gw_pipe_dpdk_flow *pipe,
 					struct doca_gw_monitor *mon, struct doca_fwd_table_cfg *cfg,
 					struct doca_gw_error *err);
 
-int doca_gw_dpdk_init_port(uint16_t port_id);
+uint16_t doca_gw_dpdk_private_data(void);
+int doca_gw_dpdk_start_port(uint16_t port_id,
+	struct doca_gw_port_cfg *cfg, struct doca_dpdk_port *port);
+
+extern void doca_dump_gw_match(struct doca_gw_match *match);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.h]: we need to make make it some compile option to help us debug. it is ok now, but generally no externs in code please.

File: examples/vnet/doca_gw_dpdk.c [hunk 7]
```diff
@@ -702,20 +699,42 @@ doca_gw_dpdk_build_rss_action(struct doca_dpdk_action_entry *entry,
...
+	action->conf = &queue->conf;
+	return 0;
+}
+
+static int doca_gw_dpdk_build_fwd_action(struct doca_dpdk_action_entry *action_entry,
+									struct doca_fwd_table_cfg *fwd_cfg)
+{
 	switch(fwd_cfg->type) {
 		case DOCA_FWD_RSS:
 			doca_gw_dpdk_build_rss_action(action_entry, fwd_cfg);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c]: I think you should test the cfg and if it is valid, save it for later and give user a handle. when user adds it in the pipeline you call it. you can put function pointer for implementation.

**Review by rbaryanai — CHANGES_REQUESTED** [review]

File: examples/vnet/gw.c (hunk 5 above)
Comment (BillZhou734) [reply examples/vnet/gw.c]: there i define fwd:"def_fwd" and "offload_fwd".
"def_fwd":  for default match packetes, will create flow. maybe some special core to handle it . 
"offload_fwd": for packets after offload.
I think this two part will be separated...

File: examples/vnet/doca_gw_dpdk.c (hunk 7 above)
Comment (BillZhou734) [reply examples/vnet/doca_gw_dpdk.c]: gw layer -> dpdk layer, i think we can check it once in gw layer,  right ?

File: examples/vnet/doca_gw_dpdk.h (hunk 6 above)
Comment (BillZhou734) [reply examples/vnet/doca_gw_dpdk.h]: todo next patch.

File: examples/vnet/doca_gw.h (hunk 4 above)
Comment (BillZhou734) [reply examples/vnet/doca_gw.h]: done

File: examples/vnet/doca_gw.c [hunk 8]
```diff
@@ -103,29 +86,41 @@ int doca_gw_rm_entry(uint16_t pipe_queue, struct doca_gw_pipelne_entry *entry)
...
+    struct doca_gw_port *port;
+
+	if (cfg == NULL) {
+		err->message = "port config is null.";
+		return NULL;
+	}
+	port = (struct doca_gw_port *) malloc(sizeof(struct doca_gw_port) + cfg->priv_data_size);
     if ( port == NULL ) {
-        return NULL;
+		err->message = "alloc gw port null.";
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.c:110]: maybe OOM

File: examples/vnet/doca_gw.c [hunk 9]
```diff
@@ -103,29 +86,41 @@ int doca_gw_rm_entry(uint16_t pipe_queue, struct doca_gw_pipelne_entry *entry)
...
-    }
+	port->cfg = *cfg;
+	switch(cfg->type) {
+		case DOCA_GW_PORT_DPDK:
+			DOCA_LOG_INFO("port is dpdk, matching port id");
+			// init all required data sturcures for port.
+			break;
+		case DOCA_GW_PORT_DPDK_BY_ID:
+			DOCA_LOG_INFO("new doca port type:dpdk port id:%s", cfg->devargs);
+			ret = doca_gw_dpdk_start_port(port->port_id, &port->cfg, (struct doca_dpdk_port*)port->user_data);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.c]: port_id? you don't have it here

File: examples/vnet/doca_gw.c (hunk 1 above)
Comment (BillZhou734) [reply examples/vnet/doca_gw.c]: done

File: examples/vnet/doca_gw.c [hunk 10]
```diff
@@ -7,6 +7,7 @@
 
 DOCA_LOG_MODULE(doca_gw)
 
+//can move those define to doca_gw.h ?
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.c:10]: no. on more complex scenario you will need a real object here,.

File: examples/vnet/doca_gw.c (hunk 2 above)
Comment (rbaryanai) [reply examples/vnet/doca_gw.c:122]: ? where do you set port_id?

File: examples/vnet/doca_gw.h [hunk 11]
```diff
@@ -93,11 +93,17 @@ enum doca_gw_port_type {
 
 struct doca_gw_port_cfg {
     enum doca_gw_port_type type;   /* mapping type of port */
-    uint16_t queues;                
+    uint16_t queues;				//only can do nb_queue rss, can use doca_fwd_table_cfg  ?
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.h:96]: yes, you are right.

File: examples/vnet/doca_gw.h [hunk 12]
```diff
@@ -164,11 +170,20 @@ struct doca_gw_actions {
...
+	DOCA_GW_AGING = (1 << 3),
+};
+
 struct doca_gw_monitor {
+    uint8_t flags;
     bool count;
     struct meter {
         uint64_t cir;
         uint64_t cbs;
+		uint64_t ebs;//need ?
```
Comment (rbaryanai) [inline examples/vnet/doca_gw.h:186]: we can take out because we don't implement it now anyway.

File: examples/vnet/doca_gw_dpdk.c [hunk 13]
```diff
@@ -252,11 +251,48 @@ static void doca_gw_dpdk_build_vxlan_flow_item(struct doca_dpdk_item_entry *entr
...
+}
+
+static int
+doca_gw_dpdk_modify_gre_key_item(struct doca_dpdk_item_entry *entry,
+						struct doca_gw_match *match)
+{
+	uint32_t *spec = &entry->item_data.gre_key.spec;
+
+	if (match->tun.gre.key)
+		*spec = rte_cpu_to_be_32(match->tun.gre.key);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:272]: user should put everything in network order. we keep all our data in network order. This is the rule.

File: examples/vnet/doca_gw_dpdk.c [hunk 14]
```diff
@@ -252,11 +251,48 @@ static void doca_gw_dpdk_build_vxlan_flow_item(struct doca_dpdk_item_entry *entr
...
+			struct doca_gw_match *match)
+{
+	struct rte_flow_item *flow_item = entry->item;
+	uint32_t *spec = &entry->item_data.gre_key.spec;
+	uint32_t *mask = &entry->item_data.gre_key.mask;
+
+	flow_item->type = RTE_FLOW_ITEM_TYPE_GRE_KEY;
+	if (!match->tun.gre.key)
+		return;
+	*spec = rte_cpu_to_be_32(match->tun.gre.key);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:286]: should be set by user in network order, we don't need to siwtch,

**Comment by BillZhou734** [comment]
close it , since too many different for the new PR
will start a new PR for it.

File: examples/vnet/doca_gw_dpdk.c [hunk 15]
```diff
@@ -252,11 +251,48 @@ static void doca_gw_dpdk_build_vxlan_flow_item(struct doca_dpdk_item_entry *entr
...
+	*mask = 0xffff;
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:287]: key is uint32_t so mask is not good

File: examples/vnet/doca_gw_dpdk.c [hunk 16]
```diff
@@ -252,11 +251,48 @@ static void doca_gw_dpdk_build_vxlan_flow_item(struct doca_dpdk_item_entry *entr
...
 	flow_item->mask = mask;
 	if (match->tun.vxlan.tun_id == UINT32_MAX) {
-		entry->flags |= DOCA_MODIFY_VLAN_ID;
+		entry->flags |= DOCA_MODIFY_VXLAN_VNI;
 		entry->modify_item = doca_gw_dpdk_modify_vxlan_item;
 	}
 }
 
+static void doca_gw_dpdk_build_gre_flow_item(struct doca_dpdk_item_entry *entry)
+{
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:260]: if we have a match on key, we need to make sure key flag is 1 on GRE header

File: examples/vnet/doca_gw_dpdk.c [hunk 17]
```diff
@@ -394,6 +430,10 @@ static int doca_gw_dpdk_build_item(struct doca_gw_match *match,
 			doca_gw_dpdk_build_vxlan_flow_item(NEXT_ITEM, match);
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:430]: there is an assumption here that GRE is always with a key, we need to make sure that when user build the pipeline he configures the vni.

File: examples/vnet/doca_gw_dpdk.c [hunk 18]
```diff
@@ -719,6 +782,126 @@ static int doca_gw_dpdk_build_fwd(struct doca_gw_pipe_dpdk_flow *pipe,
...
+
+	ret = rte_mtr_meter_profile_add
+		(port_id, meter_prof_id, &mp, &error);
+	if (ret != 0) {
+		DOCA_LOG_ERR("Port %u create Profile id %u error(%d) message: %s\n",
+			port_id, meter_prof_id, error.type,
+			error.message ? error.message : "(no stated reason)");
+		return 0;
+	}
+	return meter_prof_id++;
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:806]: meter_prof_is a naive approach, user can add and delete and you don't have a way to track what is in use and what not.

File: examples/vnet/doca_gw_dpdk.c [hunk 19]
```diff
@@ -719,6 +782,126 @@ static int doca_gw_dpdk_build_fwd(struct doca_gw_pipe_dpdk_flow *pipe,
...
+	params.action[RTE_COLOR_RED] = MTR_POLICER_ACTION_DROP;
+
+	ret = rte_mtr_create(port_id, meter_id, &params, 1, &error);
+	if (ret != 0) {
+		DOCA_LOG_ERR("Port %u create meter idx(%d) error(%d) message: %s\n",
+			port_id, meter_id, error.type,
+			error.message ? error.message : "(no stated reason)");
+		return 0;
+	}
+	return meter_id++;
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:850]: need a pool

File: examples/vnet/doca_gw_dpdk.c [hunk 20]
```diff
@@ -719,6 +782,126 @@ static int doca_gw_dpdk_build_fwd(struct doca_gw_pipe_dpdk_flow *pipe,
...
+	struct rte_mtr_error error;
+	static uint32_t meter_id = 1;
+
+	memset(&params, 0, sizeof(struct rte_mtr_params));
+	params.meter_enable = 1;
+	params.stats_mask = 0xffff;
+	params.use_prev_mtr_color = 0;
+	params.dscp_table = NULL;
+
+	/*create meter*/
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:837]: we will need to use new API when it is ready right?

File: examples/vnet/doca_gw_dpdk.c [hunk 21]
```diff
@@ -719,6 +782,126 @@ static int doca_gw_dpdk_build_fwd(struct doca_gw_pipe_dpdk_flow *pipe,
...
+		return -1;
+	action->type = RTE_FLOW_ACTION_TYPE_METER;
+	action->conf = meter_conf;
+	return 0;
+}
+
+static int
+doca_gw_dpdk_build_monitor_action(struct doca_gw_pipe_dpdk_flow *pipe, 
+							struct doca_gw_monitor *mon)
+{
```
Comment (rbaryanai) [inline examples/vnet/doca_gw_dpdk.c:892]: it is a good start, build please note that user can set different CBS/CIR for each flow on same pipe. so you will need to handle profiles and meters id inside the pipeline_entry struct,

**Review by rbaryanai — CHANGES_REQUESTED** [review]

**Pull request closed without merging**